### PR TITLE
fix 2 bugs: compiler: underscores in edges; oracle: edge ref

### PR DESCRIPTION
**Pull request by alixander**
1. edge case where creating a new edge with underscores AND common prefix was panicking
2. moving an object where the most nested reference is an edge

### CODE REVIEW

File: d2ir/d2ir.go [hunk 1]
```diff
@@ -770,7 +770,8 @@ func (m *Map) CreateEdge(eid *EdgeID, refctx *RefContext) (*Edge, error) {
 	if len(common) > 0 {
 		tmp := *refctx.Edge.Src
 		kp := &tmp
-		kp.Path = kp.Path[:len(common)]
+		underscores := countUnderscores(kp.IDA())
```
Comment (nhooyr) [inline d2ir/d2ir.go:773]: ooo true. Though this should instead be in `trimCommon`. or actually can you combine `resolveUnderscores` and `trimCommon`? This code should be in there.

File: d2oracle/edit_test.go [hunk 2]
```diff
@@ -2001,6 +2001,32 @@ c: {
 				assert.JSON(t, len(g.Objects), 3)
 			},
 		},
+		{
+			name: "underscore-connection",
+
+			text: `a: {
+  b
```
Comment (nhooyr) [inline d2oracle/edit_test.go:2008]: For the test case, I don't see how your fix applies. Wouldn't this have been the first reference anyway?

**Review by nhooyr — CHANGES_REQUESTED** [review]

**Comment by nhooyr** [comment]
> moving an object where the most nested reference is an edge

How is bug 2 related to underscores?

File: d2oracle/edit_test.go (hunk 2 above)
Comment (alixander) [reply d2oracle/edit_test.go:2008]: i guess not. idk how they're ordered

Comment (alixander) [reply d2oracle/edit_test.go:2008]: i add this line

Screen Shot 2023-02-14 at 5 06 11 PM

and get this
Screen Shot 2023-02-14 at 5 06 05 PM

Comment (nhooyr) [reply d2oracle/edit_test.go:2008]: Ohh that's cause this function is being called for `c` not `a.b` lol.

I don't think it's related to underscores though, you could remove the underscores and still reproduce this I believe.

File: d2ir/d2ir.go [hunk 3]
```diff
@@ -762,16 +759,12 @@ func (m *Map) CreateEdge(eid *EdgeID, refctx *RefContext) (*Edge, error) {
...
 	}
 
-	eid, m, err := eid.resolveUnderscores(m)
+	common, eid, m, err := eid.resolveToCommon(m)
 	if err != nil {
 		return nil, d2parser.Errorf(refctx.Edge, err.Error())
 	}
-	common, eid := eid.trimCommon()
-	if len(common) > 0 {
-		tmp := *refctx.Edge.Src
```
Comment (nhooyr) [inline d2ir/d2ir.go:771]: This was a hack to reuse `EnsureField`, common should be reverted back to `[]string`. It doesn't reflect something in the syntax. Ideally later we add a `ensureFieldIDA` that accepts `[]string` directly.

File: d2ir/d2ir.go [hunk 4]
```diff
@@ -379,7 +379,7 @@ func (eid *EdgeID) Match(eid2 *EdgeID) bool {
 	return true
 }
 
-func (eid *EdgeID) resolveUnderscores(m *Map) (*EdgeID, *Map, error) {
+func (eid *EdgeID) resolveToCommon(m *Map) (*d2ast.KeyPath, *EdgeID, *Map, error) {
```
Comment (nhooyr) [inline d2ir/d2ir.go:382]:
```suggestion
// resolve resolves both underscores and commons in eid.
// It returns the new eid, containing map adjusted for underscores and common ida.
func (eid *EdgeID) resolve(m *Map) (_ *EdgeID, _ *Map, common []string, _ error) {
```

**Review by nhooyr — CHANGES_REQUESTED** [review]

**Comment by nhooyr** [comment]
@alixander So do you agree the second bug isn't related to underscores?

**Comment by alixander** [comment]
yeah i guess not. an edge can just be the most nested. surprised this hasn't come up earlier though

File: d2ir/d2ir.go [hunk 5]
```diff
@@ -762,16 +760,12 @@ func (m *Map) CreateEdge(eid *EdgeID, refctx *RefContext) (*Edge, error) {
...
 	if err != nil {
 		return nil, d2parser.Errorf(refctx.Edge, err.Error())
 	}
-	common, eid := eid.trimCommon()
 	if len(common) > 0 {
-		tmp := *refctx.Edge.Src
-		kp := &tmp
-		kp.Path = kp.Path[:len(common)]
-		f, err := m.EnsureField(kp, nil)
+		f, err := m.EnsureField(d2ast.MakeKeyPath(common), nil)
```
Comment (nhooyr) [inline d2ir/d2ir.go:768]: lol i forgot i added MakeKeyPath at some point nice.

**Review by nhooyr — APPROVED** [review]
✅

**Pull request merged**